src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -70,589 +91,522 @@ public async ValueTask DisposeAsync()
+    private async ValueTask PublishLockedAsync(string subject, string? replyTo,  NatsPooledBufferWriter<byte> payloadBuffer, NatsPooledBufferWriter<byte>? headersBuffer, CancellationToken cancellationToken)
+    {
+        // Interlocked.Add(ref _counter.PendingMessages, 1);
+        await _semLock.WaitAsync(cancellationToken).ConfigureAwait(false);
         try
         {
-            if (_disposed)
-            {
-                throw new ObjectDisposedException(nameof(CommandWriter));
-            }
+            var payload = payloadBuffer.WrittenMemory;
+            var headers = headersBuffer?.WrittenMemory;
 
-            var success = false;
-            try
-            {
-                _protocolWriter.WriteUnsubscribe(sid, null);
-                success = true;
-            }
-            finally
+            // if (_disposed)
+            // {
+            //     throw new ObjectDisposedException(nameof(CommandWriter));
+            // }
+
+            PipeWriter bw;
+            lock (_lock)
             {
-                EnqueueCommand(success);
+                if (_pipeWriter == null)
+                    ThrowOnDisconnected();
+                bw = _pipeWriter!;
             }
-        }
-        finally
-        {
-            _semLock.Release();
-        }
 
-        return ValueTask.CompletedTask;
-    }
+            _protocolWriter.WritePublish(bw, subject, replyTo, headers, payload);
 
-    // only used for internal testing
-    internal async Task TestStallFlushAsync(TimeSpan timeSpan)
-    {
-        await _semLock.WaitAsync().ConfigureAwait(false);
+            payloadBuffer.Reset();
+            _pool.Return(payloadBuffer);
 
-        try
-        {
-            if (_flushTask is { IsCompletedSuccessfully: false })
+            if (headersBuffer != null)
             {
-                await _flushTask.ConfigureAwait(false);
+                headersBuffer.Reset();
+                _pool.Return(headersBuffer);
             }
 
-            _flushTask = Task.Delay(timeSpan);
+            await bw.FlushAsync(cancellationToken).ConfigureAwait(false);
         }
         finally
         {
             _semLock.Release();
+            // Interlocked.Add(ref _counter.PendingMessages, -1);
         }
     }
 
-    private async ValueTask ConnectStateMachineAsync(bool lockHeld, ClientOpts connectOpts, CancellationToken cancellationToken)
+    private async Task ReaderLoopAsync()
     {
-        if (!lockHeld)
-        {
-            if (!await _semLock.WaitAsync(_defaultCommandTimeout, cancellationToken).ConfigureAwait(false))
-            {
-                throw new TimeoutException();
-            }
-        }
+        // 8520 should fit into 6 packets on 1500 MTU TLS connection or 1 packet on 9000 MTU TLS connection
+        // assuming 40 bytes TCP overhead + 40 bytes TLS overhead per packet
+        const int maxSendMemoryLength = 8520;
+        var sendMemory = new Memory<byte>(new byte[maxSendMemoryLength]);
```

Comment:
Question: Does this get re-started between connections?
And if yes is there a way to possibly reuse?

Reply:
it lives as long as the `CommandWriter` object (and in turn `NatsConnection` object). On reconnect pipeline is renewed (not pooled at the moment) but the read loop stays in place.